### PR TITLE
Operation group fee estimation

### PR DESCRIPTION
**Pull request by anonymoussprocket**
- added operation group fee estimation
- improved constant definitions
- updated documentation
- fixed a bug with micheline int encoding
- added remote data packing

### CODE REVIEW

**Review by vishakh — APPROVED** [review]

File: src/chain/tezos/TezosNodeWriter.ts [hunk 1]
```diff
@@ -672,39 +780,47 @@ export namespace TezosNodeWriter {
...
+                gas += parseInt(result['consumed_gas']) || 0;
                 storageCost += parseInt(result['paid_storage_size_diff']) || 0;
+                if (result.kind === 'origination' || c['metadata']['operation_result']['allocated_destination_contract']) {
+                    storageCost += TezosConstants.EmptyAccountStorageBurn;
+                }
             }
         }
 
-        return { gas, storageCost };
+        const forgedOperationGroup = forgeOperations('BMLxA4tQjiu1PT2x3dMiijgvMTQo8AVxkPBPpdtM8hCfiyiC1jz', operations);
```
Comment (itamarreif) [inline src/chain/tezos/TezosNodeWriter.ts:803]: magic string again

File: src/chain/tezos/TezosNodeWriter.ts [hunk 2]
```diff
@@ -641,29 +695,83 @@ export namespace TezosNodeWriter {
...
+                    storageCost: storageLimitDelta + TezosConstants.StorageLimitPadding
+                }
+            );
         }
+
+        const gasLimitTotal = operationResources.map(r => r.gas).reduce((a, c) => a + c, 0);
+        const storageLimitTotal = operationResources.map(r => r.storageCost).reduce((a, c) => a + c, 0);
+        const forgedOperationGroup = forgeOperations('BMLxA4tQjiu1PT2x3dMiijgvMTQo8AVxkPBPpdtM8hCfiyiC1jz', operations);
+        const groupSize = forgedOperationGroup.length / 2 + 64; // operation group bytes + signature bytes
+        let estimatedFee = Math.ceil(gasLimitTotal / 10) + TezosConstants.BaseOperationFee + groupSize + TezosConstants.DefaultBakerVig;
```
Comment (itamarreif) [inline src/chain/tezos/TezosNodeWriter.ts:747]: maybe move these into little helper funcs?

File: src/chain/tezos/TezosNodeWriter.ts [hunk 3]
```diff
@@ -672,39 +780,47 @@ export namespace TezosNodeWriter {
...
+                    storageCost += TezosConstants.EmptyAccountStorageBurn;
+                }
             }
         }
 
-        return { gas, storageCost };
+        const forgedOperationGroup = forgeOperations('BMLxA4tQjiu1PT2x3dMiijgvMTQo8AVxkPBPpdtM8hCfiyiC1jz', operations);
+        const operationSize = forgedOperationGroup.length / 2 + 64; // operation bytes + signature bytes
+        const estimatedFee = Math.ceil(gas / 10) + TezosConstants.BaseOperationFee + operationSize + TezosConstants.DefaultBakerVig;
+        const estimatedStorageBurn = Math.ceil(storageCost * TezosConstants.StorageRate);
```
Comment (itamarreif) [inline src/chain/tezos/TezosNodeWriter.ts:806]: same as 747, maybe move into helper funcs?

File: src/chain/tezos/TezosNodeWriter.ts [hunk 4]
```diff
@@ -641,29 +695,83 @@ export namespace TezosNodeWriter {
...
+    export async function estimateOperationGroup(server: string, chainid: string, operations: Array<TezosP2PMessageTypes.StackableOperation>): Promise<any> {
+        let operationResources: { gas: number, storageCost: number }[] = [];
+
+        for (let i = 0; i < operations.length; i++) { // Estimate each operation.
+            const operation = operations[i];
+
+            // Estimate resources used in the set of prior transactions.
+            // If there were no prior transactions, set resource usage to 0.
+            let priorConsumedResources = { gas: 0, storageCost: 0 };
+            if (i > 0) {
```
[the remark this answers is not in the document]
Comment (anonymoussprocket) [reply src/chain/tezos/TezosNodeWriter.ts:722]: yes, to calculate the individual operation gas & storage limits

File: src/chain/tezos/TezosNodeWriter.ts [hunk 5]
```diff
@@ -641,29 +695,83 @@ export namespace TezosNodeWriter {
...
+                const priorTransactions = operations.slice(0, i);
+                priorConsumedResources = await estimateOperation(server, chainid, ...priorTransactions);
+            }
+
+            // Estimate resources for everything up to the current transaction. Newer transactions may depend on previous transactions, thus all transactions must be estimated.
+            const currentTransactions = operations.slice(0, i + 1);
+            const currentConsumedResources = await TezosNodeWriter.estimateOperation(server, chainid, ...currentTransactions);
```
[the remark this answers is not in the document]
Comment (anonymoussprocket) [reply src/chain/tezos/TezosNodeWriter.ts:729]: see above

File: src/chain/tezos/TezosNodeWriter.ts [hunk 6]
```diff
@@ -641,29 +695,83 @@ export namespace TezosNodeWriter {
...
+                {
+                    gas: gasLimitDelta + TezosConstants.GasLimitPadding,
+                    storageCost: storageLimitDelta + TezosConstants.StorageLimitPadding
+                }
+            );
         }
+
+        const gasLimitTotal = operationResources.map(r => r.gas).reduce((a, c) => a + c, 0);
+        const storageLimitTotal = operationResources.map(r => r.storageCost).reduce((a, c) => a + c, 0);
+        const forgedOperationGroup = forgeOperations('BMLxA4tQjiu1PT2x3dMiijgvMTQo8AVxkPBPpdtM8hCfiyiC1jz', operations);
```
[the remark this answers is not in the document]
Comment (anonymoussprocket) [reply src/chain/tezos/TezosNodeWriter.ts:745]: a canned block hash which doesn't need to be current for this process to work, it's been renamed in a subsequent commit

File: src/chain/tezos/TezosNodeWriter.ts (hunk 1 above)
Comment (anonymoussprocket) [reply src/chain/tezos/TezosNodeWriter.ts:803]: see above

File: src/chain/tezos/TezosNodeWriter.ts (hunk 2 above)
Comment (anonymoussprocket) [reply src/chain/tezos/TezosNodeWriter.ts:747]: will consider it for 508

File: src/chain/tezos/TezosNodeWriter.ts (hunk 3 above)
Comment (anonymoussprocket) [reply src/chain/tezos/TezosNodeWriter.ts:806]: see above

**Review by itamarreif — APPROVED** [review]

**Pull request merged**